### PR TITLE
Advent 2015 post suggestion: Composable pipelines improved

### PR DESCRIPTION
**Pull request by samuell**
Hi Guys, this is a suggestion for a follow-up on my composable pipelines post from february (https://blog.gopheracademy.com/composable-pipelines-pattern).

Best
// Samuel

### CODE REVIEW

**Comment by bketelsen** [comment]
Thank you for this.  We can publish it anytime you want, but it needs to be reviewed by at least one of our volunteer editors.  I'll try to make time when I land today.

**Comment by samuell** [comment]
Cool, thanks! No worries and no panic ... whenever you have time! :)

File: content/advent-2015/composable-pipelines-improvements.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+or let's just call it "composable concurrent pipelines". During the year, I have experimented more, and added some minor modifications to this pattern, which
+I describe below.
+
+## The old way
+
+The basic idea described in that earlier post was the following: Rather to do [like in the generator "pattern"](https://talks.golang.org/2012/concurrency.slide#25),
+where concurrent processes were packed into functions that returns the output channels on which their lazy-evaluated output will be provided,
+we instead store each process into structs, with each in-port and out-port (I use the term port from Flow-based programming here, though technically in Go, it
+is in this case just a channel stored in a named struct field) as separate struct fields, and the main code of the processes in a
+go-routine, to make it concurrent.
```
Comment (mattfarina) [inline content/advent-2015/composable-pipelines-improvements.md]: The entire first paragraph under _The old way_ is a single run-on sentence with numerous ideas. It's a bit hard to follow. What about breaking each thought into a sentence to make it an easier read?

File: content/advent-2015/composable-pipelines-improvements.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
++++
+
+I wrote a [post here on GopherAcademy](https://blog.gopheracademy.com/composable-pipelines-pattern) earlier this year,
+about an idea for a "framework-less" pattern for [Flow-Based Programming](http://www.jpaulmorrison.com/fbp/) style programs in Go,
+or let's just call it "composable concurrent pipelines". During the year, I have experimented more, and added some minor modifications to this pattern, which
+I describe below.
+
+## The old way
+
+The basic idea described in that earlier post was the following: Rather to do [like in the generator "pattern"](https://talks.golang.org/2012/concurrency.slide#25),
```
Comment (mattfarina) [inline content/advent-2015/composable-pipelines-improvements.md]: > post was the following: Rather

This part doesn't read right with a colon followed by a capitalized word.

File: content/advent-2015/composable-pipelines-improvements.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+proc1 := &AProcess{}
+proc2 := &AProcess{}
+// Connect
+proc2.In = proc1.OutChan()
+```
+
+## A better way
+
+What I realized, is that if we use a function to create new tasks, that pre-populates all the channel fields, we can
+just assign those fields to each other.
```
Comment (mattfarina) [inline content/advent-2015/composable-pipelines-improvements.md]: If you remove the commas and make _pre-populates all_ into _pre-populate_ this would be easier to read. Can you consider updating?

File: content/advent-2015/composable-pipelines-improvements.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+... then we would execute these `Run()` methods like so (adapting it to our example code here):
+
+```go
+proc1 := NewProcess()
+proc2 := NewProcess()
+go proc1.Run()
+go proc2.Run()
+```
+
+This suggests an elegant solution of the problem of driving a chain of go-routines from the main thread: Just skip the go
```
Comment (mattfarina) [inline content/advent-2015/composable-pipelines-improvements.md]: A couple suggests to make this sentence easier to read.

Change:
> elegant solution **of** the problem

to

> elegant solution **to** the problem

and change:
> thread: Just skip the

to

> thread. Skip the

File: content/advent-2015/composable-pipelines-improvements.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+The basic idea described in that earlier post was the following: Rather to do [like in the generator "pattern"](https://talks.golang.org/2012/concurrency.slide#25),
+where concurrent processes were packed into functions that returns the output channels on which their lazy-evaluated output will be provided,
+we instead store each process into structs, with each in-port and out-port (I use the term port from Flow-based programming here, though technically in Go, it
+is in this case just a channel stored in a named struct field) as separate struct fields, and the main code of the processes in a
+go-routine, to make it concurrent.
+
+The basic approach has the benefits that the code for routing out-ports to in-ports becomes a lot clearer
+than with the generator pattern, since each port is handled separately, even on separate lines of code.
+
+An unnecessary feat of that provided pattern though, was using methods for each out-ports, that would ensure that a channel
```
Comment (mattfarina) [inline content/advent-2015/composable-pipelines-improvements.md]: This might read a little more easily if

> An unnecessary feat of that provided pattern though, was using

was more like

> An unnecessary feat of that pattern was using

**Comment by mattfarina** [comment]
@samuell thanks for that post. This was useful to read. I made a few grammar suggestions to try and make the post easier to read.

**Comment by samuell** [comment]
@mattfarina Thanks a lot! I'll have a look at the mentioned issues tomorrow or so!

**Comment by samuell** [comment]
Thanks again for the feedback @mattfarina. I have fixed the issues you pointed at, and a few more in the same spirit that I found when reading it again.

I also felt I'd better skip the very last section as it went out to explain a quite complicated topic, that felt outside the scope of this post.

Feel free to comment if you find any more issues with the post!

**Comment by mattfarina** [comment]
The original post set for today couldn't be run (I checked with the author). So, I put this one up. It's now available at https://blog.gopheracademy.com/advent-2015/composable-pipelines-improvements/. Thanks.

**Comment by samuell** [comment]
Cool, thanks a lot!
